internal/authorization/authorizer.go

```diff
@@ -35,17 +48,44 @@ func (p *Authorizer) IsSecondFactorEnabled() bool {
 }
 
 // GetRequiredLevel retrieve the required level of authorization to access the object.
-func (p *Authorizer) GetRequiredLevel(subject Subject, object Object) Level {
+func (p Authorizer) GetRequiredLevel(subject Subject, object Object) Level {
 	logger := logging.Logger()
-	logger.Tracef("Check authorization of subject %s and url %s.", subject.String(), object.String())
 
-	for _, rule := range p.rules {
+	logger.Debugf("Check authorization of subject %s and object %s (method %s).",
+		subject.String(), object.String(), object.Method)
+
+	policy := p.getRequiredLevelFunc(logger, p.rules, subject, object)
+
+	if policy == nil {
+		logger.Debugf("No matching rule for subject %s and url %s... Applying default policy.",
+			subject.String(), object.String())
+
+		return p.defaultPolicy
+	}
+
+	return *policy
+}
+
+func getRequiredLevelFunc(_ *logrus.Logger, rules []*AccessControlRule, subject Subject, object Object) *Level {
+	for _, rule := range rules {
 		if rule.IsMatch(subject, object) {
-			return rule.Policy
+			return &rule.Policy
 		}
 	}
 
-	logger.Tracef("No matching rule for subject %s and url %s... Applying default policy.", subject.String(), object.String())
+	return nil
+}
+
+func getRequiredLevelTraceFunc(log *logrus.Logger, rules []*AccessControlRule, subject Subject, object Object) *Level {
```

Comment:
I think it's better to factorize the code instead of duplicating it. Maybe we can have a bool in the args of this internal function and call the function with true or false depending on the case. Otherwise,  by duplicating the code, we will start diverging too much and we can already see that some parts of that function are not covered by tests yet by codecov so that could be an issue in the long run.

Reply:
I'm actually really glad you got to see this before it was merged for this reason. I was really unsure if what I was doing was a good idea.

The intent behind it was purely optimization since this section of code is the most hit section in day to day activities. I'm happy to defer to you regarding if you think it's necessary. My gut was leaning towards it's not since it's a very low cost check.

Reply:
Glad that was helpful.
If I had to choose between two rather similar functions with two similar set of tests or one factorized implementation with only one extended set of tests, I'd go for the latter but it's only my opinion. I try to factorize when it makes it avoids duplicating the tests as well.

Reply:
Yeah that's what I applied in one of the later commits. I'm fine with it after bench-marking it. A 100 item ACL only takes about 12,000ns to parse now on a modern system.